flang/lib/Lower/IntrinsicCall.cpp

```diff
@@ -132,10 +142,21 @@ genExtremumval(FN func, FD funcDim, FC funcChar, mlir::Type resultType,
 
   bool absentDim = isAbsent(args[1]);
 
-  // For Maxval/MinVal, we call the type specific versions of
+  // For Maxval/MinVal/Product/Sum, we call the type specific versions of
   // Maxval/Minval because the result is scalar in the case below.
-  if (!hasCharacterResult && (absentDim || rank == 1))
-    return func(builder, loc, array, mask);
+  if (!hasCharacterResult && (absentDim || rank == 1)) {
+    auto ty = array.getType();
+    auto arrTy = fir::dyn_cast_ptrOrBoxEleTy(ty);
+    auto eleTy = arrTy.cast<fir::SequenceType>().getEleTy();
+    if (fir::isa_complex(eleTy)) {
+      auto result = builder.createTemporary(loc, eleTy);
+      func(builder, loc, array, mask, result);
+      return builder.create<fir::LoadOp>(loc, result);
+    }
+    auto resultBox = builder.create<fir::AbsentOp>(
+        loc, fir::BoxType::get(builder.getI1Type()));
+    return func(builder, loc, array, mask, resultBox);
+  }
 
   if (hasCharacterResult && (absentDim || rank == 1)) {
```

Comment:
I think you can refactor this so most of the code below here is in another function that gets tail called by a distinct handler for MAX/MIN and another for PROD/SUM.